### ISSUE TITLE
[feature request] [dataloader] Pad variable-sized tensors in default_collate

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#27617]
It's more and more frequent to deal with variable-sized inputs (with max size being determined at collation time). Currently one deals with those by writing a custom collate_fn, but it adds some boilerplate.

I propose to evaluate the idea of putting this into the default collate_fn. Some ideas:
0) frequent case is 1d sequences and 2d images
1) allow torch.stack / torch.cat deal with padding variable-sized inputs
2) deal with padding at collate_fn
3) convert default_collate to Collate class / module with __call__ interface. Then padding options can be put as arguments to the constructor.

**Comment by vadimkantorov** [comment]
An intermediate step could be torch.stack/cat supporting variable tensor sizes if an out-argument is provided and it has sufficient storage / size

**Comment by vadimkantorov** [comment]
@cpuhrsch Btw, an idea came up:
- Can regular dense tensors be converted to a nested tensor (even if it's just one element)?
- If so, default_collate could stack these individual tensors to produce a collated nested tensor (this code path would be used if inputs to torch.stack are all "nestable"/"nested")

**Comment by vadimkantorov** [comment]
If input somehow is explicitly marked as nestable/nested - maybe it should be rather safe?

At least for users with custom datasets, they would be able to e.g. use `return ..., my_3hw_image_tensor.to_nested()`, and then these could be collated in a nice image tensor (modulo whatever is current arch of torchvision datasets/abstractions - which I'm not a fan of)
